src/libraries/System.Private.Uri/src/System/UriScheme.cs

```diff
@@ -63,6 +66,25 @@ protected virtual void OnRegister(string schemeName, int defaultPort)
         //
         protected virtual void InitializeAndValidate(Uri uri, out UriFormatException? parsingError)
         {
+            if (uri._syntax is null)
+            {
+                throw new InvalidOperationException(SR.net_uri_NotAbsolute);
+            }
+
+            if (!ReferenceEquals(uri._syntax, this))
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_uri_UserDrivenParsing, uri._syntax.GetType()));
+            }
+
+            Debug.Assert(sizeof(Uri.Flags) == sizeof(ulong));
+
+            // If ParseMinimal is called multiple times this Uri instance may be corrupted, throw an exception instead
+            ulong previous = Interlocked.Or(ref Unsafe.As<Uri.Flags, ulong>(ref uri._flags), (ulong)Uri.Flags.CustomParser_ParseMinimalAlreadyCalled);
+            if (((Uri.Flags)previous & Uri.Flags.CustomParser_ParseMinimalAlreadyCalled) != 0)
+            {
+                throw new InvalidOperationException(SR.net_uri_InitializeCalledAlreadyOrTooLate);
+            }
+
             parsingError = uri.ParseMinimal();
```

Comment:
We hand out a Uri instance to custom parsers before leaving the constructor. We are exposing a protected `InitializeAndValidate` which calls into Uri's `PrivateParseMinimal`, but doesn't protect against it being called multiple times, potentially corrupting the instance.

I added checks to guard against
1. calling `InitializeAndValidate` on Uris not created by the parser
1. calling `InitializeAndValidate` on relative Uris - they don't have an associated UriParser and calling it results in NREs as `PrivateParseMinimal` assumes a parser is available.
1. calling it multiple times/outside of the actual `InitializeAndValidate` override as doing so could corrupt state